# Development of PyQuante 2
Rick Muller, rpmuller@gmail.com

I've been working on a rewrite of the PyQuante code, trying to get a more intuitive module structure and cleaner bindings of routines. The attached has some timing and other test results playing with the code. This isn't really meant as documentation or a gallery of pyquante2 features, but you can see some of them here.

You can fork the code from the [pyquante2 github site](https://github.com/rpmuller/pyquante2).

## What's new?
There are more modules, and the source code is spread between multiple directories. However, the goal is to have all generally-used modules imported from the main top-level module:

In [1]:
from pyquante2 import rhf,h2,h2o,c6h6,basisset

Many modules have special printing functions to support IPython:

In [3]:
h2o

Stoichiometry = H2O, Charge = 0, Multiplicity = 1
8 O     0.000000     0.000000     0.091686
1 H     1.422968     0.000000    -0.981210
1 H    -1.422968     0.000000    -0.981210

In [5]:
c6h6

Stoichiometry = H6C6, Charge = 0, Multiplicity = 1
6 C     1.865827     1.865827     0.000000
6 C     2.548767    -0.682940     0.000000
6 C     0.682940    -2.548767     0.000000
6 C    -1.865827    -1.865827     0.000000
6 C    -2.548767     0.682940     0.000000
6 C    -0.682940     2.548767     0.000000
1 H     3.317447     3.317447     0.000000
1 H     4.531717    -1.214270     0.000000
1 H     1.214270    -4.531717     0.000000
1 H    -3.317447    -3.317447     0.000000
1 H    -4.531717     1.214270     0.000000
1 H    -1.214270     4.531717     0.000000

We also try to have decent printing for those objects:

In [7]:
print(c6h6)

Stoichiometry = H6C6, Charge = 0, Multiplicity = 1
6 C     1.865827     1.865827     0.000000
6 C     2.548767    -0.682940     0.000000
6 C     0.682940    -2.548767     0.000000
6 C    -1.865827    -1.865827     0.000000
6 C    -2.548767     0.682940     0.000000
6 C    -0.682940     2.548767     0.000000
1 H     3.317447     3.317447     0.000000
1 H     4.531717    -1.214270     0.000000
1 H     1.214270    -4.531717     0.000000
1 H    -3.317447    -3.317447     0.000000
1 H    -4.531717     1.214270     0.000000
1 H    -1.214270     4.531717     0.000000


The solvers have decent printing as well:

In [10]:
h2

Stoichiometry = H2, Charge = 0, Multiplicity = 1
1 H     0.000000     0.000000     0.692179
1 H     0.000000     0.000000    -0.692179

In [9]:
bfs = basisset(h2)
bfs

cgbf((np.float64(0.0), np.float64(0.0), np.float64(0.6921792096923653)),(0, 0, 0),[3.42525091, 0.62391373, 0.1688554],[0.1543289707029839, 0.5353281424384733, 0.44463454202535485])
cgbf((np.float64(0.0), np.float64(0.0), np.float64(-0.6921792096923653)),(0, 0, 0),[3.42525091, 0.62391373, 0.1688554],[0.1543289707029839, 0.5353281424384733, 0.44463454202535485])

In [11]:
solver = rhf(h2,bfs)
solver

/Users/rick.muller/.cache/uv/archive-v0/cYn05xEIWAVArq0giMiWK/lib/python3.13/site-packages/IPython/core/formatters.py:429: FormatterWarning: text/html formatter returned invalid type <class 'bytes'> (expected <class 'str'>) for object: RHF Hamiltonian
Stoichiometry = H2, Charge = 0, Multiplicity = 1
1 H     0.000000     0.000000     0.692179
1 H     0.000000     0.000000    -0.692179
Basis set: sto3g, Nbf: 2
Status: Converged = False
  warnings.warn(


RHF Hamiltonian
Stoichiometry = H2, Charge = 0, Multiplicity = 1
1 H     0.000000     0.000000     0.692179
1 H     0.000000     0.000000    -0.692179
Basis set: sto3g, Nbf: 2
Status: Converged = False

In [8]:
solver.converge()
solver

RHF Hamiltonian
Stoichiometry = H2, Charge = 0, Multiplicity = 1
1 H     0.000000     0.000000     0.692176
1 H     0.000000     0.000000    -0.692176
Basis set: sto3g, Nbf: 2
Status: Converged = True
0  0.48556
1  -1.11710

In [12]:
print(solver)

RHF Hamiltonian
Stoichiometry = H2, Charge = 0, Multiplicity = 1
1 H     0.000000     0.000000     0.692179
1 H     0.000000     0.000000    -0.692179
Basis set: sto3g, Nbf: 2
Status: Converged = False


Simple orbital printing is built it:

In [13]:
from pyquante2 import lineplot_orbs,contourplot,line

In [15]:
points = line((0,0,-5),(0,0,5))
lineplot_orbs(points,solver.orbs[:,:2],bfs,
    title="Plots of bonding and antibonding orbitals of H2")

TypeError: arrays to stack must be passed as a "sequence" type such as list or tuple.

In [16]:
contourplot('yz',h2,solver.orbs[:,1],bfs,
    title="Contours of H2 antibonding orbital")

AttributeError: 'rhf' object has no attribute 'orbs'

## Iterators and convergence
The iterator throws an exception if convergence isn't achieved:

In [17]:
solver = rhf(h2,basisset(h2))

In [18]:
solver

RHF Hamiltonian
Stoichiometry = H2, Charge = 0, Multiplicity = 1
1 H     0.000000     0.000000     0.692179
1 H     0.000000     0.000000    -0.692179
Basis set: sto3g, Nbf: 2
Status: Converged = False

In [19]:
solver.converge(maxiters=1)

[np.float64(-1.117099429494621)]

In [20]:
solver.energies

[np.float64(-1.117099429494621)]

## Comparing timings and results with v1

In [21]:
import PyQuante, pyquante2

ModuleNotFoundError: No module named 'PyQuante'

Tests in the utils unit tests:

In [18]:
print map(PyQuante.pyints.fact2,[0,1,3,8,-1]), map(pyquante2.utils.fact2,[0,1,3,8,-1])
for a,b in [(5,2),(10,5)]:
    print PyQuante.pyints.binomial(a,b),pyquante2.utils.binomial(a,b)
print PyQuante.pyints.Fgamma(0,0),pyquante2.utils.Fgamma(0,0)

[1, 1, 3, 384, 1] [1, 1, 3, 384, 1]
10 10
252 252
0.999999996667 1.0


## Test one-e integrals and pgbfs
Tests in the one.py unit tests:

In [19]:
s1 = PyQuante.PGBF.PGBF(1,(0.,0.,0.))
s2 = pyquante2.pgbf(1)
print s1.overlap(s1),pyquante2.S(s2,s2)
print s1.kinetic(s1),pyquante2.T(s2,s2)
print s1.nuclear(s1,(0,0,0)),pyquante2.V(s2,s2,(0,0,0))

1.0 1.0
1.5 1.5
-1.59576911629 -1.59576912161


Confused about the incomplete gamma function at small values. It would be nice to be able to use the scipy routines as a replacement, but they don't seem to give the same results across the board.

In [20]:
from pyquante2.utils import *
from math import gamma
from scipy.special import gammainc
for x in [0.1,2,3,4]:
    print gamma(x),gammainc(x,1e-10),gammainc(x,1e10),gamm_inc(x,1e-10),gamm_inc(x,1e10)

9.51350769867 0.10511370061 1.0 0.999999999991 9.51350769867
1.0 4.99999999967e-21 1.0 4.99999999967e-21 1.0
2.0 1.66666666654e-31 1.0 3.33333333308e-31 2.0
6.0 4.16666666633e-42 1.0 2.4999999998e-41 6.0


This doesn't make any sense.

## Test cgbfs with one-e integrals:

In [21]:
from PyQuante.CGBF import CGBF
from pyquante2 import cgbf,S,T
c1 = CGBF((0,0,0),(0,0,0))
c1b = CGBF((0,0,1.0),(0,0,0))
exps,coefs = [],[]
for ex,co in [(3.4252509099999999, 0.15432897000000001),
           (0.62391373000000006, 0.53532813999999995),
           (0.16885539999999999, 0.44463454000000002)]:
    exps.append(ex)
    coefs.append(co)
    c1.add_primitive(ex,co)
    c1b.add_primitive(ex,co)
c1.normalize()
c1b.normalize()

c2 = cgbf(exps=exps,coefs=coefs)
c2b = cgbf((0,0,1.0),exps=exps,coefs=coefs)

print "Overlaps"
print c1.overlap(c1), S(c2,c2)
print c1.overlap(c1b), S(c2,c2b)

print "Kinetics"
print c1.kinetic(c1), T(c2,c2)
print c1.kinetic(c1b), T(c2,c2b)

Overlaps
1.0 1.0
0.796588300697 0.796588300697
Kinetics
0.760031883567 0.760031883567
0.383253671655 0.383253671655


# Integral routines and timing:

## Timing top level ERI routines

In [22]:
s = pyquante2.pgbf(1)
from pyquante2.ints import two,hgp
%timeit two.ERI(s,s,s,s)

1000 loops, best of 3: 347 us per loop


In [23]:
%timeit hgp.ERI_hgp(s,s,s,s)

1000 loops, best of 3: 238 us per loop


In [24]:
from pyquante2 import ctwo
%timeit ctwo.ERI(s,s,s,s)

100000 loops, best of 3: 6.54 us per loop


In [25]:
%timeit ctwo.ERI_hgp(s,s,s,s)

100000 loops, best of 3: 4.98 us per loop


Nice speedup from Cython, but I had expected more relative speedup between hgp and the normal code. First run of this (little cython optimization) gave 4.4 us from ERI, and 3.84 from hgp. Python was 162 us.

## Timing comparison for different vrr routines:

In [26]:
from pyquante2.ctwo import ERI,ERI_hgp,vrr,vrr_recursive,vrr_nonrecursive
from pyquante2.ints.hgp import vrr as pyvrr

In [27]:
zero = array([0,0,0],'d')
%timeit pyvrr(zero,1.,(0,0,0),1.,zero,1.,1., zero,1.,(0,0,0),1.,zero,1.,1.,0)

1000 loops, best of 3: 214 us per loop


In [28]:
%timeit vrr(0,0,0,1.,0,0,0,1.,0,0,0,1.,1.,0,0,0,1.,0,0,0,1.,0,0,0,1.,1.,0)

1000000 loops, best of 3: 1.67 us per loop


In [29]:
%timeit vrr_recursive(0,0,0,1.,0,0,0,1.,0,0,0,1.,1.,0,0,0,1.,0,0,0,1.,0,0,0,1.,1.,0)

1000000 loops, best of 3: 1.87 us per loop


In [30]:
%timeit vrr_nonrecursive(0,0,0,1.,0,0,0,1.,0,0,0,1.,1.,0,0,0,1.,0,0,0,1.,0,0,0,1.,1.,0)

1000000 loops, best of 3: 1.92 us per loop


Again, first time through I had expected more differences between the versions. The python version is 150 us, vrr was 1.21, and vrr_recursive was 1.26. For the added complexity of vrr (storing intermediate quantities) it almost seems worth it to go back to the recursive version.

vrr is now set to the recursive version. I can remove the nonrecursive version if needed.

## Comparison of old and new ERI code:

PGBF code:

In [31]:
from PyQuante.PGBF import coulomb
from pyquante2.ctwo import ERI_hgp, ERI

s1 = PyQuante.PGBF.PGBF(1,(0.,0.,0.))
s2 = pyquante2.pgbf(1)

s1b = PyQuante.PGBF.PGBF(1,(0.,0.,1.))
s2b = pyquante2.pgbf(1,(0,0,1))

%timeit coulomb(s1,s1,s1,s1)
%timeit ERI_hgp(s2,s2,s2,s2)
print coulomb(s1,s1,s1,s1), ERI_hgp(s2,s2,s2,s2)
print coulomb(s1,s1,s1b,s1b), ERI_hgp(s2,s2,s2b,s2b), ERI(s2,s2,s2b,s2b)
print coulomb(s1,s1b,s1,s1b), ERI_hgp(s2,s2b,s2,s2b), ERI(s2,s2b,s2,s2b)

100000 loops, best of 3: 5.22 us per loop
100000 loops, best of 3: 6.05 us per loop


1.12837916333 1.12837916333
0.842700790029 0.842700790029 0.842700790029
0.415107496037 0.415107496037 0.415107496037


The new routines run at the same speed, which isn't altogether surprising, but it's nice to know I haven't lost anything due to the Cython wrappers.

CGBF code:

In [32]:
from PyQuante.CGBF import CGBF,coulomb
from pyquante2 import cgbf
from pyquante2.ctwo import ERI_hgp

c1 = CGBF((0,0,0),(0,0,0))
c1b = CGBF((0,0,1.0),(0,0,0))
exps,coefs = [],[]
for ex,co in [(3.4252509099999999, 0.15432897000000001),
           (0.62391373000000006, 0.53532813999999995),
           (0.16885539999999999, 0.44463454000000002)]:
    exps.append(ex)
    coefs.append(co)
    c1.add_primitive(ex,co)
    c1b.add_primitive(ex,co)
c1.normalize()
c1b.normalize()

c2 = cgbf(exps=exps,coefs=coefs)
c2b = cgbf((0,0,1.0),exps=exps,coefs=coefs)

%timeit coulomb(c1,c1,c1,c1)
%timeit ERI_hgp(c2,c2,c2,c2)
%timeit ERI(c2,c2,c2,c2)
print coulomb(c1,c1,c1,c1), ERI_hgp(c2,c2,c2,c2)
print coulomb(c1,c1,c1b,c1b), ERI_hgp(c2,c2,c2b,c2b)
print coulomb(c1,c1b,c1,c1b), ERI_hgp(c2,c2b,c2,c2b)

10000 loops, best of 3: 49.4 us per loop
10000 loops, best of 3: 41.4 us per loop


1000 loops, best of 3: 1.22 ms per loop


0.774605941338 0.774605941338
0.650177460815 0.650177460815
0.455901518744 0.455901518744


# SCF examples:

## LiH, STO-3G

In [33]:
from PyQuante.Ints import getbasis,getints,getT,getV
from PyQuante.hartree_fock import rhf
from PyQuante.Molecule import Molecule

LiH = Molecule('lih',
                 [(3,( .0000000000, .0000000000, .0000000000)),
                  (1,( .0000000000, .0000000000,1.629912))],
                 units='Angstroms')
bfs = getbasis(LiH,'sto-3g')
nbf = len(bfs)
nocc,nopen = LiH.get_closedopen()
assert nopen==0
S,h,Ints = getints(bfs,LiH)
en,orbe,orbs = rhf(LiH,integrals=(S,h,Ints),verbose=True)
print "SCF completed, E = ",en 
print "S = \n",S
print "h = \n",h
print "I2 = \n",Ints[:5]

SCF completed, E =  -7.86073576968
S = 
[[ 1.          0.24113665  0.          0.          0.          0.06239818]
 [ 0.24113665  1.          0.          0.          0.          0.38780341]
 [ 0.          0.          1.          0.          0.          0.        ]
 [ 0.          0.          0.          1.          0.          0.        ]
 [ 0.          0.          0.          0.          1.          0.50716955]
 [ 0.06239818  0.38780341  0.          0.          0.50716955  1.        ]]
h = 
[[-4.73076107 -1.06286658  0.          0.         -0.01551519 -0.28307001]
 [-1.06286658 -1.39257572  0.          0.         -0.12049949 -0.67475424]
 [ 0.          0.         -1.13302238  0.          0.          0.        ]
 [ 0.          0.          0.         -1.13302238  0.          0.        ]
 [-0.01551519 -0.12049949  0.          0.         -1.22977649 -0.82314252]
 [-0.28307001 -0.67475424  0.          0.         -0.82314252 -1.43835583]]
I2 = 
array('d', [1.6803951639324308, 0.2654203637230

In [34]:
from PyQuante.TestMolecules import h2o
bfs = getbasis(h2o,'sto-3g')
nbf = len(bfs)
S,h,Ints = getints(bfs,h2o)
en,orbe,orbs = rhf(h2o,integrals=(S,h,Ints),verbose=True)
print "SCF completed, E = ",en 
print "S = \n",S
print "h = \n",h
print "I2 = \n",Ints[:5]


SCF completed, E =  -74.9598407887
S = 
[[  1.00000000e+00   2.36703937e-01   0.00000000e+00   0.00000000e+00
   -3.95628520e-18   5.58181517e-02   5.58181517e-02]
 [  2.36703937e-01   1.00000000e+00   0.00000000e+00   0.00000000e+00
   -6.93671951e-18   4.84133479e-01   4.84133479e-01]
 [  0.00000000e+00   0.00000000e+00   1.00000000e+00   0.00000000e+00
    0.00000000e+00   3.18049483e-01  -3.18049483e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00
    0.00000000e+00   0.00000000e+00   0.00000000e+00]
 [ -3.95628520e-18  -6.93671951e-18   0.00000000e+00   0.00000000e+00
    1.00000000e+00  -2.39804296e-01  -2.39804296e-01]
 [  5.58181517e-02   4.84133479e-01   3.18049483e-01   0.00000000e+00
   -2.39804296e-01   1.00000000e+00   2.54586613e-01]
 [  5.58181517e-02   4.84133479e-01  -3.18049483e-01   0.00000000e+00
   -2.39804296e-01   2.54586613e-01   1.00000000e+00]]
h = 
[[ -3.27375153e+01  -7.61673386e+00   0.00000000e+00   0.00000000e+00
    1.92563167e-

In [35]:
from pyquante2 import molecule,lih,basisset,rhf
from pyquante2.ints.integrals import onee_integrals,twoe_integrals
from pyquante2.utils import dmat,trace2,geigh

lih = molecule([
    (3,.0000000000, .0000000000, .0000000000),
    (1, .0000000000, .0000000000,1.629912)],
    units='Angstroms')

bfs = basisset(lih,'sto3g')

i1 = onee_integrals(bfs,lih)
i2 = twoe_integrals(bfs)
print "S = \n",i1.S
print "h = \n",i1.T + i1.V
print "I2 = \n",i2._2e_ints[:5]
s = rhf(lih,bfs)
s.converge()

S = 
[[ 1.          0.24113665  0.          0.          0.          0.06239931]
 [ 0.24113665  1.          0.          0.          0.          0.38780552]
 [ 0.          0.          1.          0.          0.          0.        ]
 [ 0.          0.          0.          1.          0.          0.        ]
 [ 0.          0.          0.          0.          1.          0.50717121]
 [ 0.06239931  0.38780552  0.          0.          0.50717121  1.        ]]
h = 
[[-4.73076276 -1.06286698  0.          0.         -0.01551535 -0.28307525]
 [-1.06286698 -1.39257664  0.          0.         -0.12049997 -0.67475941]
 [ 0.          0.         -1.13302317  0.          0.          0.        ]
 [ 0.          0.          0.         -1.13302317  0.          0.        ]
 [-0.01551535 -0.12049997  0.          0.         -1.22977768 -0.82314683]
 [-0.28307525 -0.67475941  0.          0.         -0.82314683 -1.4383608 ]]
I2 = 
[[[[  1.68039516e+00   2.65420364e-01   0.00000000e+00   0.00000000e+00
      0.00

[-3.1596039710068511,
 -7.8207331845299244,
 -7.8597189212209964,
 -7.8606098325514955,
 -7.860713927416132,
 -7.8607399825987567]

In [36]:
from pyquante2 import molecule,h2o,basisset,rhf
from pyquante2.ints.integrals import onee_integrals,twoe_integrals
from pyquante2.utils import dmat,trace2,geigh

bfs = basisset(h2o,'sto3g')

i1 = onee_integrals(bfs,h2o)
i2 = twoe_integrals(bfs)
print "S = \n",i1.S
print "h = \n",i1.T + i1.V
print "I2 = \n",i2._2e_ints[:5]
s = rhf(h2o,bfs)
s.converge()

S = 
[[  1.00000000e+00   2.36703937e-01   0.00000000e+00   0.00000000e+00
   -4.36556721e-18   5.58187997e-02   5.58187997e-02]
 [  2.36703937e-01   1.00000000e+00   0.00000000e+00   0.00000000e+00
    0.00000000e+00   4.84136727e-01   4.84136727e-01]
 [  0.00000000e+00   0.00000000e+00   1.00000000e+00   0.00000000e+00
    0.00000000e+00   3.18050818e-01  -3.18050818e-01]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   1.00000000e+00
    0.00000000e+00   0.00000000e+00   0.00000000e+00]
 [ -4.36556721e-18   0.00000000e+00   0.00000000e+00   0.00000000e+00
    1.00000000e+00  -2.39805302e-01  -2.39805302e-01]
 [  5.58187997e-02   4.84136727e-01   3.18050818e-01   0.00000000e+00
   -2.39805302e-01   1.00000000e+00   2.54589471e-01]
 [  5.58187997e-02   4.84136727e-01  -3.18050818e-01   0.00000000e+00
   -2.39805302e-01   2.54589471e-01   1.00000000e+00]]
h = 
[[ -3.27375212e+01  -7.61673524e+00   0.00000000e+00   0.00000000e+00
    1.92565169e-02  -1.81003520e+00  -1.81003520e+0

[-65.138374025697487,
 -74.467494698388464,
 -74.958013783771676,
 -74.959757388633875,
 -74.959839790498989,
 -74.959853191419228]

In [37]:
from pyquante2 import h2o,basisset,rhf
from pyquante2.utils import dmat
from pyquante2.ints.integrals import twoe_integrals,twoe_integrals_compressed
bfs = basisset(h2o,'sto3g')
solver = rhf(h2o,bfs)
solver.converge()
D = dmat(solver.orbs,h2o.nocc())
i2 = twoe_integrals(bfs)
i2c = twoe_integrals_compressed(bfs)

In [38]:
%timeit twoe_integrals(bfs)

10 loops, best of 3: 193 ms per loop


In [39]:
%timeit twoe_integrals_compressed(bfs)

10 loops, best of 3: 182 ms per loop


In [40]:
%timeit i2.get_j(D)

100000 loops, best of 3: 10.1 us per loop


In [41]:
%timeit i2c.get_j(D)

100 loops, best of 3: 4.86 ms per loop


In [ ]:
(i2.get_j(D)-i2c.get_j(D)).max()

2.2204460492503131e-16

There are several significant results here:

* The uncompressed integrals don't take any longer
* The get_j() function is much faster.
* The uncompressed ones are now

In [ ]:
from pyquante2 import h2o,basisset,rhf
from pyquante2.utils import dmat
from pyquante2.ints.integrals import twoe_integrals,twoe_integrals_compressed
bfs = basisset(h2o,'6-31g**')
solver = rhf(h2o,bfs)
solver.converge()
D = dmat(solver.orbs,h2o.nocc())
i2 = twoe_integrals(bfs)
i2c = twoe_integrals_compressed(bfs)